### PR TITLE
Update/remove unnecessary queries

### LINKED ISSUE #3676: Optimization: Unnecessary queries with get_option
Jetpack is making unnecessary queries with each page refresh trying to get options that do not exist in database. All queries are triggered by get_option with this option_name and return 0 rows:

WHERE option_name = 'embed_autourls'
WHERE option_name = 'jetpack_sync_non_public_post_stati'
WHERE option_name = 'jetpack_sso_remove_login_form'
WHERE option_name = 'jetpack_testimonial'
WHERE option_name = 'jetpack_relatedposts'
WHERE option_name = 'gravatar_disable_hovercards'
WHERE option_name = 'sharedaddy_disable_resources'
WHERE option_name = 'safecss_revision_migrated'
WHERE option_name = 'safecss'

+ 1 SELECT to posts

SELECT c5gh4s_posts.ID
FROM c5gh4s_posts 
WHERE 1=1 
AND c5gh4s_posts.post_type = 'safecss'
AND ((c5gh4s_posts.post_status = 'publish')) 
ORDER BY c5gh4s_posts.post_date DESC
LIMIT 0, 1

Some of those queries, won't be triggered if related module will be disabled/enabled. For example if I disable "related posts" the query "WHERE option_name = 'jetpack_relatedposts'" won't be called. Re-enabling "Related posts" will make this query to show up again. If this option is needed on each page, I think it should be added to database and autoloaded.

Topic first started here: https://wordpress.org/support/topic/fewer-queries-with-each-page-refresh?replies=2

### PR DESCRIPTION
**Pull request by jubstuff**
This is a work in progress, as an attempt to fix #3676 

#### Changes proposed in this Pull Request:

Many of the queries are triggered on every page only because the options don't exist yet in the database. 
When the option is set, it is autoloaded. So it makes sense to set defaults for these.

##### Affected options

 * `jetpack-twitter-cards-site-tag`
 * `open_graph_protocol_site_type`
 * `facebook_admins`
 * `safecss_add`
 * `safecss_revision_migrated`
 * `safecss`
 * `jetpack_portfolio`
 * `jetpack_testimonial`
 * `gravatar_disable_hovercards`
 * `sync_non_public_post_stati`
 * `relatedposts`
 * `sharedaddy_disable_resources`
 * `jetpack_sso_remove_login_form`
 * `verification_services_codes`
 * `embed_autourls`

The only exception is the `embed_autourls` option, that is replaced with `true`, as it's no longer used by WordPress (https://core.trac.wordpress.org/changeset/22070)

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If [Grunt](http://gruntjs.com/) is installed on your testing environment, run `grunt jshint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [x] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [ ] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

### CODE REVIEW

File: modules/shortcodes/polldaddy.php [hunk 1]
```diff
@@ -573,7 +573,7 @@ function polldaddy_link( $content ) {
 	add_filter( 'the_content_rss', 'polldaddy_link', 1 );
 
 	/** This filter is documented in modules/shortcodes/youtube.php */
-	if ( apply_filters( 'jetpack_comments_allow_oembed', get_option( 'embed_autourls' ) ) ) {
+	if ( apply_filters( 'jetpack_comments_allow_oembed', true ) ) {
```
Comment (mdawaffe) [inline modules/shortcodes/polldaddy.php]: I think we should stick with the `get_option( 'embed_autourls' )` call. The option may no longer exist, but people could still be attaching things to the `default_option_embed_autourls` filter (which would be a bit silly, but it's possible).

**Comment by mdawaffe** [comment]
For the `false === get_option( $option_name )`, I'd rather pass a default parameter that we know can never be stored. Something like: `'does-not-exist' === get_option( $option_name, 'does-not-exist' )`.

That way, there can't ever be confusion about `false` being returned because there was nothing stored v. because `false` was stored.

Also, when this PR stores `0`, should it instead store `false`?

**Comment by mdawaffe** [comment]
What happens when a site has one of these modules already active then upgrades Jetpack to the newest version? Will the `jetpack_activate_module_` filter run for that already active module? If not, the site won't ever get this improvement.

File: modules/shortcodes/polldaddy.php (hunk 1 above)
Comment (jubstuff) [inline modules/shortcodes/polldaddy.php]: I'm not against sticking with the `get_option`, but it seems that WordPress is already [attaching a filter to that](https://github.com/WordPress/WordPress/blob/master/wp-includes/default-filters.php#L333). Do you think there could be filters with a higher priority attached?

**Comment by jubstuff** [comment]
Hello @mdawaffe,
I'll reply inline

> That way, there can't ever be confusion about false being returned because there was nothing stored v. because false was stored.
Also, when this PR stores 0, should it instead store false?

Actually I don't think (or better, I don't know :) ) if it's possible to store `false` using `add_option`/`update_option` as it will result in an empty string in the former case, and no value saved at all in the latter. 
That's because the `option_value` field is a `longtext`, I suppose, but I could be wrong.

That's why I chose `false` as the parameter to test if the option is alredy saved in the option table or not.

>What happens when a site has one of these modules already active then upgrades Jetpack to the newest version? Will the jetpack_activate_module_ filter run for that already active module? If not, the site won't ever get this improvement.

I'm working on that :+1:

**Comment by jubstuff** [comment]
I have added the update routine so that these options get set when upgrading. 
I've tested it going from 3.9.6 and 4.0.2: it seems to be working well :)

File: modules/shortcodes/polldaddy.php (hunk 1 above)
Comment (mdawaffe) [inline modules/shortcodes/polldaddy.php]: I think I understand this a bit better now. WordPress didn't remove the option, exactly, it removed the UI.

WordPress uses the `default_option_` filter (as opposed to the `pre_option_` or `option_` filters).  That means if someone has the option set to some value in the database, `get_option( 'embed_autourls' )` will return that value (for example, `false`).

So removing the option call breaks backward compatibility in two ways:
1. People could be using one of the options filters to adjust this behavior, and
2. People could have a value set in the database to adjust this behavior.

I'm not against breaking backward compatibility here - I just want to make sure we know that we're doing it.  It's probably very rare that anyone is depending on either of those behaviors.

**Comment by mdawaffe** [comment]
> That's why I chose `false` as the parameter to test if the option is alredy saved in the option table or not.

That makes sense - thanks for the explanation.  There is a weird edge case:

```php
var_dump( get_option( 'new', 'nope' ) ); // "nope"
add_option( 'new', false );
var_dump( get_option( 'new', 'nope' ) ); // false
```

Within the same pageload where the option is set, `get_option()` returns `false` instead of the empty string :) (I did not know that until now.)

`false` is still probably a reasonable choice. What do you think? Also, `false` is the default value for the second parameter, so we don't need to pass it explicitly. Thoughts?

**Comment by mdawaffe** [comment]
> I have added the update routine so that these options get set when upgrading.  I've tested it going from 3.9.6 and 4.0.2: it seems to be working well :)

Makes sense to me, thanks.

**Comment by jubstuff** [comment]
> There is a weird edge case

Thank you for the edge case: I didn't know about it either :smile: 

> `false` is still probably a reasonable choice. What do you think? Also, `false` is the default value

Well, I think too that `false` should be safe enough. 
Also, about the default value, I just like to be very explicit about what I should be expecting from `get_option`, but I can remove it :+1:

File: modules/shortcodes/polldaddy.php (hunk 1 above)
Comment (jubstuff) [inline modules/shortcodes/polldaddy.php]: Thank you for the explanation, @mdawaffe 

I didn't find many references to this option online, that's why I thought that it could be safe enough to remove the query completely.

In any case, If we want to maintain backward compatibility, we could use the same pattern as for the other options, and set a default value. What do you think?

Comment (mdawaffe) [inline modules/shortcodes/polldaddy.php]: > we could use the same pattern as for the other options, and set a default value

Then every site would have an option set that WordPress doesn't even need anymore :)

I've changed my opinion from my original stance. I now think the current code in this PR (removing the call to `get_option()`) is fine even though it breaks backward compatibility.

Unless @jeherve, @kraftbj, @samhotchkiss, or some other Jetpack person disagrees, let's stick with your original idea.

**Comment by mdawaffe** [comment]
> I just like to be very explicit about what I should be expecting.

Makes sense.

**Comment by mdawaffe** [comment]
@jubstuff is there anything left to do on this PR or is it ready to merge?

**Comment by jubstuff** [comment]
@mdawaffe It is ready for me :+1:

**Comment by jeherve** [comment]
Would you mind rebasing, so we can take a look and see about including this in 4.2, scheduled to be released in a few weeks?

Thanks!

**Comment by jubstuff** [comment]
@jeherve Done! ;)

**Comment by jeherve** [comment]
@gravityrail Could you review this PR, since you tackled this issue in #4281 a few days ago?

Thanks!

**Comment by gravityrail** [comment]
@jeherve @jubstuff this PR was superceded by #4281, and any future PR's which do "forced autoloading" should probably follow the patterns established in that change, i.e. calling `Jetpack_Options::get_option_and_ensure_autoload()`.

Closing this for now.
